File: trunk/CharasLike/Classes/Images.vb
Public Class RPGImage
    Private RawBitmap_ As Bitmap
    Public Property RawBitmap() As Bitmap
        Get
            Return RawBitmap_
        End Get
        Set(ByVal value As Bitmap)
            RawBitmap_ = value
        End Set
    End Property

    Private OriginalBitmap_ As Bitmap
    Public Property OriginalBitmap() As Bitmap
        Get
            Return OriginalBitmap_
        End Get
        Set(ByVal value As Bitmap)
            OriginalBitmap_ = value
        End Set
    End Property

    Private Box_ As RPGPictureBox
    Public Property Box() As RPGPictureBox
        Get
            Return Box_
        End Get
        Set(ByVal value As RPGPictureBox)
            Box_ = value
        End Set
    End Property

    Private Path_ As String
    Public Property Path() As String
        Get
            Return Path_
        End Get
        Set(ByVal value As String)
            Path_ = value
        End Set
    End Property

    Private Name_ As String
    Public Property Name() As String
        Get
            Return Name_
        End Get
        Set(ByVal value As String)
            Name_ = value
        End Set
    End Property

    Private GameFile_ As RPGGameFile
    Public Property GameFile() As RPGGameFile
        Get
            Return GameFile_
        End Get
        Set(ByVal value As RPGGameFile)
            GameFile_ = value
        End Set
    End Property
End Class

Public Class GameNodeEntry
    Private Subdirs_ As New Collections.ObjectModel.Collection(Of GameNodeEntry)
    Public ReadOnly Property Subdirs() As Collections.ObjectModel.Collection(Of GameNodeEntry)
        Get
            Return Subdirs_
        End Get
    End Property

    Private Images_ As New Collections.ObjectModel.Collection(Of RPGImage)
    Public ReadOnly Property Images() As Collections.ObjectModel.Collection(Of RPGImage)
        Get
            Return Images_
        End Get
    End Property

    Private NodeName_ As String
    Public Property NodeName() As String
        Get
            Return NodeName_
        End Get
        Set(ByVal value As String)
            NodeName_ = value
        End Set
    End Property

    Private Path_ As String
    Public Property Path() As String
        Get
            Return Path_
        End Get
        Set(ByVal value As String)
            Path_ = value
        End Set
    End Property

    Private GameFile_ As RPGGameFile
    Public Property GameFile() As RPGGameFile
        Get
            Return GameFile_
        End Get
        Set(ByVal value As RPGGameFile)
            GameFile_ = value
        End Set
    End Property

    Public Sub RecurseNodeAndAddAll()
        For Each Image In Images
            ImagePacker.FlowLayoutPanel2.Controls.Add(Image.Box.CreatePictureBoxCopy())
        Next

        For Each Subdir In Subdirs
            Subdir.RecurseNodeAndAddAll()
        Next
    End Sub
End Class

Public Class GameNodeNode
    Inherits TreeNode
    Private RPGNode_ As GameNodeEntry
    Public Property RPGNode() As GameNodeEntry
        Get
            Return RPGNode_
        End Get
        Set(ByVal value As GameNodeEntry)
            RPGNode_ = value
        End Set
    End Property
End Class

Public Class GameModeMenuItem
    Inherits ToolStripMenuItem
    Private GameFile_ As RPGGameFile
    Public Property GameFile() As RPGGameFile
        Get
            Return GameFile_
        End Get
        Set(ByVal value As RPGGameFile)
            GameFile_ = value
        End Set
    End Property
End Class

Public Class Images
    Public Shared RPGNodes As New Collections.ObjectModel.Collection(Of GameNodeEntry)
    Public Shared RPGGames As New Collections.ObjectModel.Collection(Of RPGGameFile)

    Shared Function CompareGameFiles(ByVal Left As RPGGameFile, ByVal Right As RPGGameFile) As Boolean
        If Left Is Nothing Or Right Is Nothing Then Return False

        If Left.SheetRows <> Right.SheetRows Then Return False
        If Left.SheetColumns <> Right.SheetColumns Then Return False
        If Left.PreviewFrame <> Right.PreviewFrame Then Return False
        If Left.Is8Bit <> Right.Is8Bit Then Return False
        If Left.GameName <> Right.GameName Then Return False
        If Left.Animations.Count <> Right.Animations.Count Then Return False

        For i As Integer = 0 To Left.Animations.Count - 1
            If Left.Animations(i).Frames.Count <> Right.Animations(i).Frames.Count Then Return False

            For z As Integer = 0 To Left.Animations(i).Frames.Count - 1
                If Left.Animations(i).Frames(z) <> Right.Animations(i).Frames(z) Then Return False
            Next
        Next

        Return True
    End Function

    Shared Function FindGame(ByVal CheckGame As RPGGameFile) As RPGGameFile
        If CheckGame Is Nothing Then Return Nothing

        For Each GameFile In RPGGames
            If String.Compare(GameFile.FilePath, CheckGame.FilePath, True) = 0 Then
                ' Check game file
                If CompareGameFiles(CheckGame, GameFile) = True Then
                    Return GameFile ' Gotcha
                Else
                    If MsgBox("The required game file (" + CheckGame.FilePath + ") varies from your game file version. In order for this to work properly, you must replace your game file with this one." + vbNewLine + vbNewLine + "A backup will be made at " + CheckGame.FilePath + ".bak. Continue?", MsgBoxStyle.OkCancel) = MsgBoxResult.Ok Then
                        ' FIXME: do it
                        IO.File.Copy(GameFile.FilePath, GameFile.FilePath + ".bak")

                        Using File As New IO.BinaryWriter(IO.File.Open(GameFile.FilePath, IO.FileMode.Truncate))
                            CheckGame.Save(File)
                        End Using

                        Return GameFile
                    Else
                        Return Nothing
                    End If
                End If
            End If
        Next
        Return Nothing
    End Function

    Shared Function FindGameByPath(ByVal Path As String) As RPGGameFile
        For Each GameFile In RPGGames
            If String.Compare(GameFile.Path, Path, True) = 0 Then
                Return GameFile
            End If
        Next
        Return Nothing
    End Function

    Shared Function SearchImage(ByVal List As Collections.ObjectModel.Collection(Of GameNodeEntry), ByVal Paths As String(), ByVal CurIndex As Integer) As RPGImage
        For Each Node In List
            If String.Compare(Node.NodeName, Paths(CurIndex), True) = 0 Then
                CurIndex += 1

                ' Are we at the end of the list now?
                If CurIndex = Paths.Length - 1 Then
                    ' See if we have said image
                    For Each Img In Node.Images
                        If String.Compare(Img.Name, Paths(CurIndex), True) = 0 Then
                            Return Img
                        End If
                    Next
                    Return Nothing
                Else
                    ' Keep searching from here
                    Return SearchImage(Node.Subdirs, Paths, CurIndex)
                End If
            End If
        Next

        Return Nothing
    End Function

    Shared Function FindImage(ByVal Path As String) As RPGImage
        Dim Main As String = Path.Substring(5)
        Dim Paths() As String = Main.Split(New String() {"\"}, StringSplitOptions.None)

        Return SearchImage(RPGNodes, Paths, 0)
    End Function

    Shared Function BitmapCompare(ByVal Bmp1 As Bitmap, ByVal Bmp2 As Bitmap) As Boolean
        If Bmp1.Size <> Bmp2.Size Then Return False

        Using fpBmp1 As New FastPixel(Bmp1, True), fpBmp2 As New FastPixel(Bmp2, True)
            For y As Integer = 0 To fpBmp1.Height - 1
                For x As Integer = 0 To fpBmp1.Width - 1
                    If fpBmp1.GetPixel(x, y) <> fpBmp2.GetPixel(x, y) Then Return False
                Next
            Next
        End Using

        Return True ' Same
    End Function

    Shared Function Convert8BPPTo32BPP(ByVal Convert As Image) As Bitmap
        If Convert Is Nothing Then Return Nothing

        Dim TempMap = New Bitmap(Convert.Width, Convert.Height, Imaging.PixelFormat.Format32bppArgb)
        Dim Gfx As Graphics = Graphics.FromImage(TempMap)
        Gfx.DrawImage(Convert, New Rectangle(0, 0, Convert.Width, Convert.Height))
        Gfx.Dispose()

        Return TempMap
    End Function

    Shared Sub LoadRPGImage(ByVal Node As GameNodeEntry, ByVal Path As String)
        If Node Is Nothing Or Path Is Nothing Then Return

        Dim Img As New RPGImage
        Using TmpImage As New Bitmap(Path)
            Dim TempMap As Bitmap = Convert8BPPTo32BPP(TmpImage)

            Img.RawBitmap = New Bitmap(TempMap.Width, TempMap.Height, Imaging.PixelFormat.Format32bppArgb)
            Img.OriginalBitmap = New Bitmap(TmpImage)
            Img.GameFile = Node.GameFile

            Using ImgFP As New FastPixel(Img.RawBitmap, True)
                Using TempMapFP As New FastPixel(TempMap, True)
                    Dim TransColor As Color = TempMapFP.GetPixel(0, 0)

                    For i As Integer = 0 To TempMap.Width - 1
                        For z As Integer = 0 To TempMap.Height - 1
                            If (TempMapFP.GetPixel(i, z) = TransColor) Then
                                ImgFP.SetPixel(i, z, Color.FromArgb(0, 0, 0, 0))
                            Else
                                ImgFP.SetPixel(i, z, TempMapFP.GetPixel(i, z))
                            End If
                        Next
                    Next
                End Using
            End Using
        End Using

        Img.Path = Path
        Img.Name = Path.Substring(Path.LastIndexOf("\") + 1)
        Img.Box = RPGPictureBox.Create(Img)

        Node.Images.Add(Img)
    End Sub

    Shared Sub AddAndRecurseNodes(ByVal Nodes As TreeNodeCollection, ByVal GameNode As GameNodeEntry)
        If Nodes Is Nothing Or GameNode Is Nothing Then Return

        ' Insert me
        Dim Node As New GameNodeNode
        Node.RPGNode = GameNode
        Node.Text = GameNode.NodeName

        Nodes.Add(Node)

        For Each SubNode In GameNode.Subdirs
            AddAndRecurseNodes(Node.Nodes, SubNode)
        Next
    End Sub

    Shared Sub RecurseSubdirectories(ByVal Node As GameNodeEntry)
        If Node Is Nothing Then Return

        Dim entries As String() = IO.Directory.GetDirectories(Node.Path)

        For Each directory In entries
            If directory.Substring(directory.LastIndexOf("\") + 1).StartsWith(".") Then Continue For

            Dim Entry As New GameNodeEntry
            Entry.NodeName = directory.Substring(directory.LastIndexOf("\") + 1)
            Entry.Path = directory
            Entry.GameFile = CurrentGameFile

            Dim Files As New List(Of String)
            Files.AddRange(IO.Directory.GetFiles(directory, "*.png"))
            Files.AddRange(IO.Directory.GetFiles(directory, "*.bmp"))
            Files.AddRange(IO.Directory.GetFiles(directory, "*.jpg"))

            ' See if path has any images
            For Each Img In Files
                LoadRPGImage(Entry, Img)
            Next

            RecurseSubdirectories(Entry)

            Node.Subdirs.Add(Entry)
        Next
    End Sub

    Public Shared CurrentGameFile As RPGGameFile

    Shared Sub GameMenuItemClick(ByVal sender As Object, ByVal e As EventArgs)
        Dim Item As GameModeMenuItem = CType(sender, GameModeMenuItem)

        If Item.GameFile Is CurrentGameFile Then Return

        ' Set game
        CurrentGameFile.MenuItem.Checked = False
        CurrentGameFile = Item.GameFile
        Item.Checked = True

        My.Settings.CurrentGame = CurrentGameFile.Path
    End Sub

    Shared Sub ClearImagesSubDirs(ByVal Node As GameNodeEntry)
        If Node Is Nothing Then Return

        For Each SubNode In Node.Subdirs
            SubNode.Images.Clear()
            ClearImagesSubDirs(SubNode)
        Next
    End Sub

    Shared Sub ClearImages()
        RPGNodes.Clear()
        ItemsWindow.FlowLayoutPanel1.Controls.Clear()
        ItemsWindow.FlowLayoutPanel1.Controls.Add(RPGPictureBox.EmptyBox)
        ItemsWindow.TreeView1.Nodes.Clear()
    End Sub

    ' Loads images and game files
    Shared Sub LoadImages()
        Dim subdirectoryEntries As String() = IO.Directory.GetDirectories("data\")

        Dim WillBeCurrentGame As RPGGameFile = Nothing
        For Each Game In subdirectoryEntries
            If Game.Substring(Game.LastIndexOf("\") + 1).StartsWith(".") Then Continue For
            Dim fileEntries As String() = IO.Directory.GetFiles(Game)

            Dim Entry As New GameNodeEntry
            Entry.NodeName = Game.Substring(Game.LastIndexOf("\") + 1)
            Entry.Path = Game

            ' Is there a .gam file in here?
            For Each File In fileEntries
                If File.EndsWith(".gam") Then
                    Entry.GameFile = New RPGGameFile
                    Using infile = IO.File.Open(File, IO.FileMode.Open)
                        Using Reader As New IO.BinaryReader(infile)
                            Entry.GameFile.FilePath = File

                            If Reader IsNot Nothing Then
                                Entry.GameFile.Load(Reader)
                                Entry.GameFile.Path = Entry.Path
                            End If

                            RPGGames.Add(Entry.GameFile)

                            CurrentGameFile = Entry.GameFile
                            If My.Settings.CurrentGame = CurrentGameFile.Path Then WillBeCurrentGame = CurrentGameFile
                            Exit For
                        End Using
                    End Using
                End If
            Next

            RecurseSubdirectories(Entry)

            RPGNodes.Add(Entry)
        Next

        If WillBeCurrentGame IsNot Nothing Then
            CurrentGameFile = WillBeCurrentGame
        Else
            CurrentGameFile = Nothing
        End If

        ' Fill in the menu
        For Each Game In RPGGames
            Dim MenuItem As GameModeMenuItem = New GameModeMenuItem
            MenuItem.GameFile = Game
            Game.MenuItem = MenuItem
            MenuItem.Text = Game.GameName
            AddHandler MenuItem.Click, AddressOf GameMenuItemClick

            If CurrentGameFile IsNot Nothing Then
                If Game Is CurrentGameFile Then MenuItem.Checked = True
            ElseIf My.Settings.FirstRun Or My.Settings.CurrentGame = "" Or CurrentGameFile Is Nothing Then
                CurrentGameFile = Game
                My.Settings.CurrentGame = CurrentGameFile.Path
                MenuItem.Checked = True
            End If

            Canvas.SetGameMenuItem.DropDownItems.Add(MenuItem)
        Next

        ' Fill the tree view in ItemsWindow
        For Each Node In RPGNodes
            AddAndRecurseNodes(ItemsWindow.TreeView1.Nodes, Node)
        Next
    End Sub

    ' Call this when you want to refresh the image list.
    Shared Sub ReloadImages()
        Dim Backup As String = CurrentGameFile.Path

        ' Back up all image data in the layers
        ' FIXME: do all characters
        For Each Layer In CharacterSelect.CharacterList.CurrentCharacter().Character.Layers
            If Layer.Image IsNot Nothing Then
                Dim Img As RPGImage = Layer.Image

                Layer.Image = New RPGImage
                Layer.Image.Path = Img.Path
            End If
            Layer.ReloadImagesSubLayers(False)
        Next

        ' Clear nodes
        ClearImages()

        ' Clear games
        Canvas.SetGameMenuItem.DropDownItems.Clear()
        RPGGames.Clear()

        LoadImages()

        ' Reset game
        CurrentGameFile = FindGameByPath(Backup)

        ' Fix layer data
        ' FIXME2: above
        For Each Layer In CharacterSelect.CharacterList.CurrentCharacter().Character.Layers
            If Layer.Image IsNot Nothing Then
                Layer.Image = FindImage(Layer.Image.Path)
                Layer.UpdateFlippedImage()
            End If
            Layer.ReloadImagesSubLayers(True)
        Next

        ' Re-draw incase we changed
        Canvas.UpdateDrawing()
    End Sub
End Class

File: Classes/Images.cs
using System.Collections.Generic;
using System;
using System.Drawing;
using System.Diagnostics;
using System.Collections;
using System.Windows.Forms;
using System.IO;

namespace OpenCharas
{
	public class RPGImage
	{
		public Bitmap RawBitmap { get; set; }
		public Bitmap OriginalBitmap { get; set; }
		public RPGPictureBox Box { get; set; }
		public string Path { get; set; }
		public string Name { get; set; }
		public RPGGameFile GameFile { get; set; }
	}

	public class GameNodeEntry
	{
		public List<GameNodeEntry> Subdirs { get; private set; }
		public List<RPGImage> Images { get; private set; }
		public string NodeName { get; set; }
		public string Path { get; set; }
		public RPGGameFile GameFile { get; set; }

		public GameNodeEntry()
		{
			Subdirs = new List<GameNodeEntry>();
			Images = new List<RPGImage>();
		}

		public void RecurseNodeAndAddAll()
		{
			//foreach (var Image in Images)
				//Program.imagePackerForm.FlowLayoutPanel2.Controls.Add(Image.Box.CreatePictureBoxCopy());

			foreach (var Subdir in Subdirs)
				Subdir.RecurseNodeAndAddAll();
		}
	}

	public class GameNodeNode : TreeNode
	{
		public GameNodeEntry RPGNode { get; set; }
	}

	public class GameModeMenuItem : ToolStripMenuItem
	{
		public RPGGameFile GameFile { get; set; }
	}

	public static class Images
	{
		public static List<GameNodeEntry> RPGNodes = new List<GameNodeEntry>();
		public static List<RPGGameFile> RPGGames = new List<RPGGameFile>();

		static public bool CompareGameFiles(RPGGameFile Left, RPGGameFile Right)
		{
			if (Left == null || Right == null)
				return false;

			if (Left.SheetRows != Right.SheetRows)
				return false;
			if (Left.SheetColumns != Right.SheetColumns)
				return false;
			if (Left.PreviewFrame != Right.PreviewFrame)
				return false;
			if (Left.Is8Bit != Right.Is8Bit)
				return false;
			if (Left.GameName != Right.GameName)
				return false;
			if (Left.Animations.Count != Right.Animations.Count)
				return false;

			for (int i = 0; i <= Left.Animations.Count - 1; i++)
			{
				if (Left.Animations[i].Frames.Count != Right.Animations[i].Frames.Count)
					return false;

				for (int z = 0; z <= Left.Animations[i].Frames.Count - 1; z++)
				{
					if (Left.Animations[i].Frames[z] != Right.Animations[i].Frames[z])
						return false;
				}
			}

			return true;
		}

		static public RPGGameFile FindGame(RPGGameFile CheckGame)
		{
			if (CheckGame == null)
			{
				return null;
			}

			foreach (var GameFile in RPGGames)
			{
				if (string.Compare(GameFile.FilePath, CheckGame.FilePath, true) == 0)
				{
					// Check game file
					if (CompareGameFiles(CheckGame, GameFile) == true)
						return GameFile; // Gotcha
					else
					{
						if (MessageBox.Show("The required game file (" + CheckGame.FilePath + ") varies from your game file version. In order for this to work properly, you must replace your game file with this one.\n\nA backup will be made at " + CheckGame.FilePath + ".bak. Continue?", "", MessageBoxButtons.OKCancel) == DialogResult.OK)
						{
							// FIXME: do it
							System.IO.File.Copy(GameFile.FilePath, GameFile.FilePath + ".bak");

							using (System.IO.BinaryWriter File = new System.IO.BinaryWriter(System.IO.File.Open(GameFile.FilePath, System.IO.FileMode.Truncate)))
								CheckGame.Save(File);

							return GameFile;
						}
						else
							return null;
					}
				}
			}
			return null;
		}

		static public RPGGameFile FindGameByPath(string Path)
		{
			foreach (var GameFile in RPGGames)
			{
				if (string.Compare(GameFile.Path, Path, true) == 0)
					return GameFile;
			}
			return null;
		}

		static public RPGImage SearchImage(List<GameNodeEntry> List, string[] Paths, int CurIndex)
		{
			foreach (var Node in List)
			{
				if (string.Compare((string)Node.NodeName, Paths[CurIndex], true) == 0)
				{
					CurIndex++;

					// Are we at the end of the list now
					if (CurIndex == Paths.Length - 1)
					{
						// See if we have said image
						foreach (var Img in Node.Images)
						{
							if (string.Compare((string)Img.Name, Paths[CurIndex], true) == 0)
								return Img;
						}
						return null;
					}
					else
					{
						// Keep searching from here
						return SearchImage(Node.Subdirs, Paths, CurIndex);
					}
				}
			}

			return null;
		}

		static public RPGImage FindImage(string Path)
		{
			string Main = Path.Substring(5);
			string[] Paths = Main.Split('\\'.ToString().ToCharArray());

			return SearchImage(RPGNodes, Paths, 0);
		}

		static public bool BitmapCompare(Bitmap Bmp1, Bitmap Bmp2)
		{
			if (Bmp1.Size != Bmp2.Size)
			{
				return false;
			}

			using (FastPixel fpBmp1 = new FastPixel(Bmp1, true))
			{
				using (FastPixel fpBmp2 = new FastPixel(Bmp2, true))
				{
					for (int y = 0; y <= fpBmp1.Height - 1; y++)
					{
						for (int x = 0; x <= fpBmp1.Width - 1; x++)
						{
							if (fpBmp1.GetPixel(x, y) != fpBmp2.GetPixel(x, y))
							{
								return false;
							}
						}
					}
				}
			}


			return true; // Same
		}

		static public Bitmap Convert8BPPTo32BPP(Image Convert)
		{
			if (Convert == null)
			{
				return null;
			}

			var TempMap = new Bitmap(Convert.Width, Convert.Height, System.Drawing.Imaging.PixelFormat.Format32bppArgb);
			Graphics Gfx = Graphics.FromImage(TempMap);
			Gfx.DrawImage(Convert, new Rectangle(0, 0, Convert.Width, Convert.Height));
			Gfx.Dispose();

			return TempMap;
		}

		static public void LoadRPGImage(GameNodeEntry Node, string Path)
		{
			if (Node == null || Path == null)
			{
				return;
			}

			RPGImage Img = new RPGImage();
			using (Bitmap TmpImage = new Bitmap(Path))
			{
				Bitmap TempMap = Convert8BPPTo32BPP(TmpImage);

				Img.RawBitmap = new Bitmap(TempMap.Width, TempMap.Height, System.Drawing.Imaging.PixelFormat.Format32bppArgb);
				Img.OriginalBitmap = new Bitmap(TmpImage);
				Img.GameFile = Node.GameFile;

				using (FastPixel ImgFP = new FastPixel(Img.RawBitmap, true))
				{
					using (FastPixel TempMapFP = new FastPixel(TempMap, true))
					{
						Color TransColor = TempMapFP.GetPixel(0, 0);

						for (int i = 0; i <= TempMap.Width - 1; i++)
						{
							for (int z = 0; z <= TempMap.Height - 1; z++)
							{
								if (TempMapFP.GetPixel(i, z) == TransColor)
								{
									ImgFP.SetPixel(i, z, Color.FromArgb(0, 0, 0, 0));
								}
								else
								{
									ImgFP.SetPixel(i, z, TempMapFP.GetPixel(i, z));
								}
							}
						}
					}

				}

			}


			Img.Path = Path;
			Img.Name = Path.Substring((int)(Path.LastIndexOf("\\") + 1));
			Img.Box = RPGPictureBox.Create(Img);

			Node.Images.Add(Img);
		}

		static public void AddAndRecurseNodes(TreeNodeCollection Nodes, GameNodeEntry GameNode)
		{
			if (Nodes == null || GameNode == null)
			{
				return;
			}

			// Insert me
			GameNodeNode Node = new GameNodeNode();
			Node.RPGNode = GameNode;
			Node.Text = GameNode.NodeName;

			Nodes.Add(Node);

			foreach (var SubNode in GameNode.Subdirs)
			{
				AddAndRecurseNodes(Node.Nodes, SubNode);
			}
		}

		static public void RecurseSubdirectories(GameNodeEntry Node)
		{
			if (Node == null)
			{
				return;
			}

			string[] entries = System.IO.Directory.GetDirectories(Node.Path);

			foreach (var directory in entries)
			{
				if (directory.Substring(directory.LastIndexOf("\\") + 1).StartsWith("."))
				{
					continue;
				}

				GameNodeEntry Entry = new GameNodeEntry();
				Entry.NodeName = (string)(directory.Substring(directory.LastIndexOf("\\") + 1));
				Entry.Path = directory.ToString();
				Entry.GameFile = CurrentGameFile;

				List<string> Files = new List<string>();
				Files.AddRange(System.IO.Directory.GetFiles(directory.ToString(), "*.png"));
				Files.AddRange(System.IO.Directory.GetFiles(directory.ToString(), "*.bmp"));
				Files.AddRange(System.IO.Directory.GetFiles(directory.ToString(), "*.jpg"));

				// See if path has any images
				foreach (var Img in Files)
				{
					LoadRPGImage(Entry, Img.ToString());
				}

				RecurseSubdirectories(Entry);

				Node.Subdirs.Add(Entry);
			}
		}

		public static RPGGameFile CurrentGameFile;

		static public void GameMenuItemClick(object sender, EventArgs e)
		{
			GameModeMenuItem Item = (GameModeMenuItem)sender;

			if (Item.GameFile == CurrentGameFile)
				return;

			if (!Program.canvasForm.NewSheet())
				return;

			// Set game
			CurrentGameFile.MenuItem.Checked = false;
			CurrentGameFile = Item.GameFile;
			Item.Checked = true;

			Canvas.Settings.CurrentGame = CurrentGameFile.Path;
		}

		static public void ClearImagesSubDirs(GameNodeEntry Node)
		{
			if (Node == null)
				return;

			foreach (var SubNode in Node.Subdirs)
			{
				SubNode.Images.Clear();
				ClearImagesSubDirs(SubNode);
			}
		}

		static public void ClearImages()
		{
			RPGNodes.Clear();
			Program.itemsWindowForm.imageViewerContainer1.Images = null;
			Program.itemsWindowForm.TreeView1.Nodes.Clear();
		}

		// Loads images and game files
		static public void LoadImages()
		{
			string[] subdirectoryEntries = System.IO.Directory.GetDirectories("data\\");

			RPGGameFile WillBeCurrentGame = null;
			foreach (var Game in subdirectoryEntries)
			{
				if (Game.Substring(Game.LastIndexOf("\\") + 1).StartsWith("."))
				{
					continue;
				}
				string[] fileEntries = System.IO.Directory.GetFiles(Game.ToString());

				GameNodeEntry Entry = new GameNodeEntry();
				Entry.NodeName = (string)(Game.Substring(Game.LastIndexOf("\\") + 1));
				Entry.Path = Game.ToString();

				bool hasGame = false;

				// Is there a .gam file in here
				foreach (var File in fileEntries)
				{
					if (File.EndsWith(".gam"))
					{
						Entry.GameFile = new RPGGameFile();
						using (var infile = System.IO.File.Open(File.ToString(), System.IO.FileMode.Open))
						{
							using (System.IO.BinaryReader Reader = new System.IO.BinaryReader(infile))
							{
								Entry.GameFile.FilePath = File.ToString();

								if (Reader != null)
								{
									Entry.GameFile.Load(Reader);
									Entry.GameFile.Path = Entry.Path;
								}

								RPGGames.Add(Entry.GameFile);

								CurrentGameFile = Entry.GameFile;
								if (Canvas.Settings.CurrentGame == CurrentGameFile.Path)
									WillBeCurrentGame = CurrentGameFile;

								hasGame = true;
								break;
							}
						}
					}
				}

				if (hasGame)
				{
					RecurseSubdirectories(Entry);
					RPGNodes.Add(Entry);
				}
			}

			if (WillBeCurrentGame != null)
			{
				CurrentGameFile = WillBeCurrentGame;
			}
			else
			{
				CurrentGameFile = null;
			}

			// Fill in the menu
			foreach (var Game in RPGGames)
			{
				GameModeMenuItem MenuItem = new GameModeMenuItem();
				MenuItem.GameFile = Game;
				Game.MenuItem = MenuItem;
				MenuItem.Text = (string)Game.GameName;
				MenuItem.Click += new System.EventHandler(GameMenuItemClick);

				if (CurrentGameFile != null)
				{
					if (Game == CurrentGameFile)
					{
						MenuItem.Checked = true;
					}
				}
				else if (Canvas.Settings.FirstRun || Canvas.Settings.CurrentGame == "" || CurrentGameFile == null)
				{
					CurrentGameFile = Game;
					Canvas.Settings.CurrentGame = CurrentGameFile.Path;
					MenuItem.Checked = true;
				}

				Program.canvasForm.SetGameMenuItem.DropDownItems.Add(MenuItem);
			}

			// Fill the tree view in ItemsWindow
			foreach (var Node in RPGNodes)
			{
				AddAndRecurseNodes(Program.itemsWindowForm.TreeView1.Nodes, Node);
			}
		}

		// Call this when you want to refresh the image list.
		static public void ReloadImages()
		{
			string Backup = CurrentGameFile.Path;

			// Back up all image data in the layers
			// FIXME: do all characters
			foreach (RPGCharacterRowNode Row in Program.characterSelectForm.TreeView1.Nodes)
			{
				foreach (RPGCharacterNode Col in Row.Nodes)
				{
					foreach (var Layer in Col.Character.Layers)
					{
						if (Layer.Image != null)
						{
							RPGImage Img = Layer.Image;

							Layer.Image = new RPGImage();
							Layer.Image.Path = Img.Path;
						}
						Layer.ReloadImagesSubLayers(false);
					}
				}
			}

			// Clear nodes
			ClearImages();

			// Clear games
			Program.canvasForm.SetGameMenuItem.DropDownItems.Clear();
			RPGGames.Clear();

			LoadImages();

			// Reset game
			CurrentGameFile = FindGameByPath(Backup);

			// Fix layer data
			// FIXME2: above
			foreach (RPGCharacterRowNode Row in Program.characterSelectForm.TreeView1.Nodes)
			{
				foreach (RPGCharacterNode Col in Row.Nodes)
				{
					foreach (var Layer in Col.Character.Layers)
					{
						if (Layer.Image != null)
						{
							Layer.Image = FindImage(Layer.Image.Path);
							Layer.UpdateFlippedImage();
						}
						Layer.ReloadImagesSubLayers(true);
					}
				}
			}

			// Re-draw incase we changed
			Program.canvasForm.UpdateDrawing();
		}
	}

	public static class Extensions
	{
		public static void CopyTo(this Stream from, Stream to, int bufferSize = 4096)
		{
			long fromPos = 0, toPos = 0;

			if (from.CanSeek)
				fromPos = from.Position;
			if (to.CanSeek)
				toPos = to.Position;

			byte[] buffer = new byte[bufferSize];
			while (true)
			{
				int bytes = from.Read(buffer, 0, bufferSize);

				if (bytes == 0)
					break;

				to.Write(buffer, 0, bytes);
			}

			if (from.CanSeek)
				from.Position = fromPos;
			if (to.CanSeek)
				to.Position = toPos;
		}
		/// <summary>
		/// Converts an image into an icon.
		/// </summary>
		/// <param name="img">The image that shall become an icon</param>
		/// <param name="size">The width and height of the icon. Standard
		/// sizes are 16x16, 32x32, 48x48, 64x64.</param>
		/// <param name="keepAspectRatio">Whether the image should be squashed into a
		/// square or whether whitespace should be put around it.</param>
		/// <returns>An icon!!</returns>
		public static Icon ToIcon(this Image img, int size, bool keepAspectRatio)
		{
			Bitmap square = new Bitmap(size, size); // create new bitmap
			Graphics g = Graphics.FromImage(square); // allow drawing to it

			int x, y, w, h; // dimensions for new image

			if (!keepAspectRatio || img.Height == img.Width)
			{
				// just fill the square
				x = y = 0; // set x and y to 0
				w = h = size; // set width and height to size
			}
			else
			{
				// work out the aspect ratio
				float r = (float)img.Width / (float)img.Height;

				// set dimensions accordingly to fit inside size^2 square
				if (r > 1)
				{ // w is bigger, so divide h by r
					w = size;
					h = (int)((float)size / r);
					x = 0; y = (size - h) / 2; // center the image
				}
				else
				{ // h is bigger, so multiply w by r
					w = (int)((float)size * r);
					h = size;
					y = 0; x = (size - w) / 2; // center the image
				}
			}

			// make the image shrink nicely by using HighQualityBicubic mode
			g.InterpolationMode = System.Drawing.Drawing2D.InterpolationMode.HighQualityBicubic;
			g.DrawImage(img, x, y, w, h); // draw image with specified dimensions
			g.Flush(); // make sure all drawing operations complete before we get the icon

			// following line would work directly on any image, but then
			// it wouldn't look as nice.
			return Icon.FromHandle(square.GetHicon());
		}
	}
}
